datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -858,6 +889,57 @@ impl SMJStream {
         }
     }
 
+    fn free_reservation(&mut self, buffered_batch: BufferedBatch) -> Result<()> {
+        // Shrink memory usage for in memory batches only
+        if buffered_batch.spill_file.is_none() && buffered_batch.batch.is_some() {
+            self.reservation
+                .try_shrink(buffered_batch.size_estimation)?;
+        }
```

Comment:
We should also handle `else` cases, i.e., spilling file is `Some` and `batch` is also `Some`, and both are `None`, etc.

Reply:
I think those cases are not possible but the current code doesn't make that clear

Here is a proposal that I think makes it clearer what states are possible: https://github.com/comphead/arrow-datafusion/pull/297